paddle/fluid/framework/tensor_extension.h

```diff
@@ -0,0 +1,496 @@
+  if (count) {
+    for (; count && b != e; ++b, --count) {
+      f(*b);
+    }
+  } else {
+    for (; b != e; ++b) {
+      f(*b);
+    }
+  }
+}
+
+template <class U, class TensorType>
+TensorOutStreamer<U, TensorType>::TensorOutStreamer(const char* _label,
+                                                    const char* _name,
+                                                    const TensorType& _tensor,
+                                                    DataLayout _layout)
+    : label(_label), name(_name), tensor(_tensor), limit(0), layout(_layout) {}
+
+template <class U, class TensorType>
+TensorOutStreamer<U, TensorType>& TensorOutStreamer<U, TensorType>::setLimit(
+    size_t _limit) {
+  limit = _limit;
+  return *this;
+}
+
+template <class U, class TensorType>
+const U* TensorOutStreamer<U, TensorType>::begin() const {
+  // return tensor.data<U>();
+  return reinterpret_cast<const U*>(tensor.raw_data());
+}
+
+template <class U, class TensorType>
+const U* TensorOutStreamer<U, TensorType>::end() const {
+  return begin() + tensor.memory_size() / sizeof(U);
+}
+
+template <class U, class TensorType>
+int TensorOutStreamer<U, TensorType>::checksum() const {
+  boost::crc_32_type result;
+  result.process_bytes(reinterpret_cast<const unsigned char*>(begin()),
+                       tensor.memory_size());
+  return result.checksum();
+}
+
+template <class U>
+struct type_desc;
+
+template <>
+struct type_desc<signed char> {
+  enum { pad = 2, break_line = 256 };
+  static const char* name() { return "signed_char"; }
+  static std::ostream& format(std::ostream& out, signed char v) {
+    out << std::setw(pad) << std::setfill('0') << std::hex
+        << (static_cast<int>(v) & 0xFF);
+    return out;
+  }
+};
+
+template <>
+struct type_desc<unsigned char> {
+  enum { pad = 2, break_line = 256 };
+  static const char* name() { return "unsigned_char"; }
+  static std::ostream& format(std::ostream& out, unsigned char v) {
+    out << std::setw(pad) << std::setfill('0') << std::hex
+        << (static_cast<int>(v) & 0xFF);
+    return out;
+  }
+};
+
+template <>
+struct type_desc<float> {
+  enum { pad = 8, break_line = 16 };
+  static const char* name() { return "float"; }
+  static std::ostream& format(std::ostream& out, float v) {
+    out << std::setw(pad) << std::setfill(' ') << v << " ";
+    return out;
+  }
+};
+
+template <class U, class TensorType>
+std::ostream& operator<<(std::ostream& out,
+                         const TensorOutStreamer<U, TensorType>& ts) {
+  auto& tensor = ts.tensor;
+  auto& dims = tensor.dims();
+  auto& conf = TensorDumpConfig<>::get();
+
+  out << std::setw(8) << std::setfill(' ') << (TensorDumpConfig<>::NextRecord())
+      << ") type=[" << type_desc<U>::name() << "]  => label=" << ts.label
+      << " name=" << ((ts.name) ? ts.name : "undef") << " crc32=" << std::hex
+      << ts.checksum() << std::dec << " elem=" << tensor.numel()
+      << " dims=" << dims.size() << "=>";
```

Comment:
Let's make it more readable:
```suggestion
  out << std::setw(8) << std::setfill(' ') 
      << (TensorDumpConfig<>::NextRecord())
      << ") type=[" << type_desc<U>::name() 
      << "]  => label=" << ts.label
      << " name=" << ((ts.name) ? ts.name : "undef") 
      << " crc32=" << std::hex << ts.checksum() << std::dec 
      << " elem=" << tensor.numel()
      << " dims=" << dims.size() << "=>";
```